x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -282,33 +305,133 @@ public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId,
+                () -> {
+                    logger.trace("{} background renewal of retention lease during restore", shardId);
+                    final ThreadContext threadContext = threadPool.getThreadContext();
+                    try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                        // we have to execute under the system context so that if security is enabled the renewal is authorized
+                        threadContext.markAsSystemContext();
+                        renewRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
+                    } catch (final Exception e) {
+                        assert e instanceof ElasticsearchSecurityException == false : e;
+                        logger.warn(new ParameterizedMessage("{} background renewal of retention lease failed during restore", shardId), e);
+                        throw e;
+                    }
+                },
+                RETENTION_LEASE_RENEW_INTERVAL_SETTING.get(indexShard.indexSettings().getSettings()),
+                Ccr.CCR_THREAD_POOL_NAME);
+
+        // TODO: There should be some local timeout. And if the remote cluster returns an unknown session
+        //  response, we should be able to retry by creating a new session.
+        try (RestoreSession restoreSession = openSession(metadata.name(), remoteClient, leaderShardId, indexShard, recoveryState)) {
+            restoreSession.restoreFiles();
+            updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, indexShard.routingEntry().index());
+        } catch (Exception e) {
+            throw new IndexShardRestoreFailedException(indexShard.shardId(), "failed to restore snapshot [" + snapshotId + "]", e);
+        } finally {
+            logger.trace("{} canceling background renewal of retention lease at the end of restore", shardId);
+            renewable.cancel();
+        }
+    }
+
+    private void createEmptyStore(final IndexShard indexShard, final ShardId shardId) {
         final Store store = indexShard.store();
         store.incRef();
         try {
             store.createEmpty(indexShard.indexSettings().getIndexMetaData().getCreationVersion().luceneVersion);
-        } catch (EngineException | IOException e) {
+        } catch (final EngineException | IOException e) {
             throw new IndexShardRecoveryException(shardId, "failed to create empty store", e);
         } finally {
             store.decRef();
         }
+    }
 
-        Map<String, String> ccrMetaData = indexShard.indexSettings().getIndexMetaData().getCustomData(Ccr.CCR_CUSTOM_METADATA_KEY);
-        String leaderUUID = ccrMetaData.get(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY);
-        Index leaderIndex = new Index(shardId.getIndexName(), leaderUUID);
-        ShardId leaderShardId = new ShardId(leaderIndex, shardId.getId());
+    void acquireRetentionLeaseOnLeader(
```

Comment:
I hope we can move this to `CcrRetentionLeases.java ` and share it with ShardFollowTask but this is blocking vs non-blocking in ShardFollowTask.

Reply:
That is the plan indeed. 😇